### PR TITLE
Migrate android_device to null safety

### PR DESCRIPTION
**Pull request by jmagman**
Migrate android_device and android_device_discovery.

Part of #71511.

### CODE REVIEW

**Review by christopherfujino — APPROVED** [review]
some nits, but LGTM

File: packages/flutter_tools/lib/src/android/android_device.dart [hunk 1]
```diff
@@ -90,60 +88,53 @@ class AndroidDevice extends Device {
...
+  Future<String?> _getProperty(String name) async {
+    return (await _properties)[name];
   }
 
   @override
-  Future<bool> get isLocalEmulator async {
-    if (_isLocalEmulator == null) {
-      final String hardware = await _getProperty('ro.hardware');
+  late final Future<bool> isLocalEmulator = () async {
+      final String? hardware = await _getProperty('ro.hardware');
```
[the remark this answers is not in the document]
Comment (jmagman) [reply packages/flutter_tools/lib/src/android/android_device.dart:127]: Nope, fixed.

File: packages/flutter_tools/lib/src/android/android_device.dart [hunk 2]
```diff
@@ -754,11 +742,16 @@ class AndroidDevice extends Device {
 
   @override
   Future<MemoryInfo> queryMemoryInfo() async {
+    final AndroidApk? package = _package;
+    if (package == null) {
+      _logger.printTrace('Android package unknown, skipping dumpsys meminfo.');
```
[the remark this answers is not in the document]
Comment (jmagman) [reply packages/flutter_tools/lib/src/android/android_device.dart:747]: Sure, done.

File: packages/flutter_tools/lib/src/android/android_device.dart [hunk 3]
```diff
@@ -1128,34 +1125,34 @@ class AdbLogReader extends DeviceLogReader {
...
+        final Match? fatalMatch = _tombstoneLine.firstMatch(line);
 
         if (fatalMatch != null) {
           acceptLine = true;
 
-          line = fatalMatch[1];
+          line = fatalMatch[1]!;
 
-          if (_tombstoneTerminator.hasMatch(fatalMatch[1])) {
+          if (_tombstoneTerminator.hasMatch(fatalMatch[1]!)) {
```
[the remark this answers is not in the document]
Comment (jmagman) [reply packages/flutter_tools/lib/src/android/android_device.dart:1150]: Thanks for the catch.

**Pull request merged**